File: python/twelve-days/twelve_days.py
def recite(start_verse, end_verse):
    return  [verse(number, number)[0] for number in range(start_verse, end_verse + 1)]

def verse(start_verse, end_verse):
    GIFTS = ["a Partridge in a Pear Tree.",
             "two Turtle Doves",
             "three French Hens",
             "four Calling Birds",
             "five Gold Rings",
             "six Geese-a-Laying",
             "seven Swans-a-Swimming",
             "eight Maids-a-Milking",
             "nine Ladies Dancing",
             "ten Lords-a-Leaping",
             "eleven Pipers Piping",
             "twelve Drummers Drumming",
             ]
    NUMBER_DAYS = ["first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"]

    return [f"On the {NUMBER_DAYS[start_verse - 1]} day of Christmas my true love gave to me: "
            f"{(', ').join(GIFTS[end_verse - 1:0:-1])}{', and ' if (start_verse > 1) else ''}{GIFTS[0]}"]

File: ruby/twelve-days/twelve_days.rb
class TwelveDays
  NUMBER_DAYS = %w(first second third fourth fifth sixth seventh eighth ninth tenth eleventh twelfth)
  GIFTS = ['a Partridge in a Pear Tree',
           'two Turtle Doves',
           'three French Hens',
           'four Calling Birds',
           'five Gold Rings',
           'six Geese-a-Laying',
           'seven Swans-a-Swimming',
           'eight Maids-a-Milking',
           'nine Ladies Dancing',
           'ten Lords-a-Leaping',
           'eleven Pipers Piping',
           'twelve Drummers Drumming'
  ]

  def self.recite(first_couplet, last_couplet)
    (first_couplet..last_couplet).each_with_object([]) { |couplet_number, result| result << new.verse(couplet_number) }.flatten
  end

  def self.song
    (1..12).map { |strophe_number| self.recite(strophe_number, strophe_number) }.join("\n\n") + "\n"
  end

  def verse(couplet_number)
    couplet_number -= 1
    initial = "On the #{NUMBER_DAYS[couplet_number]} day of Christmas my true love gave to me: "
    final = ''
    strophe = GIFTS[1..couplet_number].reverse.join(', ')
    final += ', and ' unless strophe.empty?
    final += GIFTS.first + '.'
    [initial + strophe + final]
  end
end
